stripe/src/main/java/com/stripe/android/CustomerSession.java

```diff
@@ -0,0 +1,299 @@
+            return;
+        }
+
+        Runnable fetchCustomerRunnable = new Runnable() {
+            @Override
+            public void run() {
+                Customer customer = retrieveCustomerWithKey(key, mStripeApiProxy, nowCalendar);
+                Message message = mUiThreadHandler.obtainMessage(CUSTOMER_RETRIEVED, customer);
+                mUiThreadHandler.sendMessage(message);
+            }
+        };
+
+        executeRunnable(fetchCustomerRunnable);
+    }
+
+    void updateKeyIfNecessary(@NonNull final EphemeralKey key, @Nullable Calendar nowCalendar) {
+        if (!isTimeInPast(key.getExpires(), nowCalendar)) {
+            return;
+        }
+
+        mKeyProvider.fetchEphemeralKey(
+                StripeApiHandler.API_VERSION,
+                new CustomerKeyUpdateListener(this));
+    }
+
+    void executeRunnable(@NonNull Runnable runnable) {
+
+        // In automation, run on the main thread.
+        if (mStripeApiProxy != null) {
+            runnable.run();
+            return;
+        }
+
+        mThreadPoolExecutor.execute(runnable);
+    }
+
+    void onKeyUpdated(String rawKey) {
+        EphemeralKey key = EphemeralKey.fromString(rawKey);
+        mEphemeralKey = key == null ? EphemeralKey.getEmptyKey() : key;
+        updateCustomerIfNecessary(mEphemeralKey, mProxyNowCalendar);
+    }
+
+    void onKeyError(int code, @Nullable String errorMessage) {
+        // Not yet handling this case
+    }
+
+    private void initializeMainThreadHandler() {
+        mUiThreadHandler = new Handler(Looper.getMainLooper()) {
+            @Override
+            public void handleMessage(Message msg) {
+                super.handleMessage(msg);
+
+                Object messageObject = msg.obj;
+
+                switch (msg.what) {
+                    case CUSTOMER_RETRIEVED:
+                        if (messageObject instanceof Customer) {
+                            mCustomer = (Customer) messageObject;
+                        }
+                        break;
+                }
+            }
+        };
+    }
+
+    private void initializeThreadPoolExecutor() {
+        mThreadPoolExecutor = new ThreadPoolExecutor(
+                THREAD_POOL_SIZE,
+                THREAD_POOL_SIZE,
+                KEEP_ALIVE_TIME,
+                KEEP_ALIVE_TIME_UNIT,
+                mNetworkQueue);
+    }
+
+    /**
+     * Calls the Stripe API (or a test proxy) to fetch a customer. If the provided key is expired,
+     * this method <b>does not</b> update the key.
+     * Use {@link #updateCustomerIfNecessary(EphemeralKey, Calendar)} to validate the key
+     * before refreshing the customer.
+     *
+     * @param key the {@link EphemeralKey} used for this access
+     * @param proxy a {@link StripeApiProxy} to intercept calls to the real servers
+     * @return a {@link Customer} if one can be found with this key, or {@code null} if one cannot.
+     */
+    @Nullable
+    static Customer retrieveCustomerWithKey(
```

Comment:
Also, this can't be a non-static method because it invokes the static method from the API class.